File: merge_recent_branches.py
#!/usr/bin/env python3
"""
Systematic merge script for recent cursor branches
"""

import subprocess
import json
import sys
from datetime import datetime, timedelta

def run_git(cmd, check=True, timeout=60):
    """Run a git command"""
    full_cmd = f"git {cmd}"
    print(f"→ {full_cmd}")
    try:
        result = subprocess.run(
            full_cmd,
            shell=True,
            capture_output=True,
            text=True,
            timeout=timeout
        )
        if result.stdout:
            print(result.stdout)
        if result.stderr and result.returncode != 0:
            print(f"Error: {result.stderr}")
        return result.returncode == 0, result.stdout.strip()
    except subprocess.TimeoutExpired:
        print(f"⚠ Command timed out")
        return False, ""
    except Exception as e:
        print(f"⚠ Exception: {e}")
        return False, ""

def get_recent_branches():
    """Get recent cursor branches"""
    success, output = run_git("for-each-ref --sort=-committerdate refs/remotes/origin/cursor --format='%(committerdate:short) %(refname:short)'")
    if not success:
        return []
    
    branches = []
    for line in output.split('\n'):
        if line.strip():
            parts = line.split(' ', 1)
            if len(parts) == 2:
                date_str, branch = parts
                # Only include branches from today or yesterday
                try:
                    branch_date = datetime.strptime(date_str, '%Y-%m-%d')
                    if branch_date >= datetime.now() - timedelta(days=1):
                        branches.append(branch)
                except:
                    continue
    
    return branches[:50]  # Limit to 50 most recent

def safe_merge(branch_name):
    """Safely merge a branch"""
    print(f"\n🔄 Attempting to merge {branch_name}...")
    
    # Check if branch has changes not in main
    success, diff = run_git(f"diff main..{branch_name} --name-only", check=False)
    if not success or not diff.strip():
        print(f"  → No changes to merge, skipping")
        return True
    
    # Try to merge
    success, output = run_git(f"merge {branch_name} --no-ff -m 'Merge {branch_name}: Recent improvements'", check=False)
    
    if success:
        print(f"  ✅ Successfully merged {branch_name}")
        return True
    else:
        # Check for conflicts
        success, conflicts = run_git("diff --name-only --diff-filter=U", check=False)
        if conflicts:
            print(f"  ⚠ Conflicts detected in: {conflicts.split()[0] if conflicts else 'unknown'}")
            print(f"  → Attempting to resolve conflicts...")
            
            # Try to resolve conflicts automatically
            try:
                # Reset to main and try merge with strategy
                run_git("reset --hard HEAD", check=False)
                success, _ = run_git(f"merge {branch_name} -X theirs --no-ff -m 'Auto-merge {branch_name}: Resolved conflicts'", check=False)
                
                if success:
                    print(f"  ✅ Conflicts resolved and merged {branch_name}")
                    return True
                else:
                    print(f"  ❌ Could not resolve conflicts for {branch_name}")
                    run_git("reset --hard HEAD", check=False)
                    return False
            except:
                print(f"  ❌ Failed to resolve conflicts for {branch_name}")
                run_git("reset --hard HEAD", check=False)
                return False
        else:
            print(f"  ❌ Merge failed for {branch_name}: {output}")
            return False

def main():
    print("="*80)
    print(" SYSTEMATIC MERGE OF RECENT CURSOR BRANCHES")
    print("="*80)
    print()
    
    # Ensure we're on main and up to date
    print("STEP 1: Updating main branch...")
    print("-"*80)
    run_git("checkout main", check=False)
    run_git("pull origin main", check=False)
    print()
    
    # Get recent branches
    print("STEP 2: Finding recent branches...")
    print("-"*80)
    branches = get_recent_branches()
    print(f"Found {len(branches)} recent branches to process")
    print()
    
    # Merge branches
    print("STEP 3: Merging branches...")
    print("-"*80)
    
    success_count = 0
    fail_count = 0
    merge_log = []
    
    for i, branch in enumerate(branches, 1):
        print(f"\n[{i}/{len(branches)}] Processing {branch}...")
        
        # Skip if it's already merged
        success, merged = run_git(f"branch --merged main", check=False)
        if success and branch in merged:
            print(f"  → Already merged, skipping")
            continue
        
        success = safe_merge(branch)
        if success:
            success_count += 1
            merge_log.append({"branch": branch, "status": "success", "timestamp": datetime.now().isoformat()})
        else:
            fail_count += 1
            merge_log.append({"branch": branch, "status": "failed", "timestamp": datetime.now().isoformat()})
    
    # Push changes
    print(f"\nSTEP 4: Pushing changes...")
    print("-"*80)
    success, _ = run_git("push origin main", check=False)
    if success:
        print("✅ Successfully pushed to origin/main")
    else:
        print("⚠ Push failed or had issues")
    
    # Create summary
    summary = {
        "timestamp": datetime.now().isoformat(),
        "total_branches": len(branches),
        "successful_merges": success_count,
        "failed_merges": fail_count,
        "merge_log": merge_log
    }
    
    with open('merge_summary.json', 'w') as f:
        json.dump(summary, f, indent=2)
    
    print(f"\n" + "="*80)
    print(" MERGE SUMMARY")
    print("="*80)
    print(f"Total branches processed: {len(branches)}")
    print(f"Successful merges: {success_count}")
    print(f"Failed merges: {fail_count}")
    print(f"Summary saved to: merge_summary.json")
    print("="*80)

if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        print("\n\nInterrupted by user")
        sys.exit(1)
    except Exception as e:
        print(f"\n\nFatal error: {e}")
        sys.exit(1)

File: merge-recent-branches.js
#!/usr/bin/env node;
/**
 * Merge Recent Branches - Merges recent cursor and other important branches;
 */ import { execSync } from 'child_process';
import fs from 'fs';
<<<<<<< HEAD

// //Step 1: Ensure we're on main and up to date,
=======
// //Step 1: Ensure we're on main and up to date;
>>>>>>> cursor/fix-errors-and-merge-to-main-6ce7
// try {,
  execSync('git checkout main', { stdio: 'inherit' });
  execSync('git pull origin main', { stdio: 'inherit' });
//   } catch (error) {
<<<<<<< HEAD
    //   process.exit(1)
  }

//Step 2: Get recent branches,
// const recentBranches = [,
// //Step,
  1: Ensure we're on main and up to date,
// try {/* TODO: Fix JSX expression */}
=======
//   process.exit(1)};
//Step 2: Get recent branches;
// const recentBranches = [
,
];
// //Step,
  1: Ensure we're on main and up to date;
// try {/* TODO: Fix JSX expression */};
>>>>>>> cursor/fix-errors-and-merge-to-main-6ce7
  o: 'inherit' });
  execSync('git pull origin main', {/* TODO: Fix JSX expression */})
  o: 'inherit' });
//   } catch (error) {/* TODO: Fix JSX expression */};
};
//Step,
  2: Get recent branches,
// const recentBranches = [
];
  'cursor/fix-errors-and-merge-to-main-3b04',
  'cursor/fix-errors-and-merge-to-main-4800',
  'cursor/fix-errors-and-merge-to-main-9 caa',
  'cursor/fix-errors-and-merge-to-main-b8 f9',
  'cursor/fix-errors-and-merge-to-main-cefe',
  'cursor/fix-errors-and-merge-to-main-dc65',
  'cursor/fix-errors-and-merge-to-main-fbf8',
  'cursor/fix-errors-and-merge-to-main-fcbc',
  'cursor/fix-errors-and-merge-to-main-ffaa',
  'cursor/fix-errors-and-merge-to-main-ffab',
  'cursor/fix-errors-and-merge-to-main-ffee',
  'cursor/fix-errors-and-merge-to-main-5696',
  'cursor/fix-errors-and-merge-to-main-6 ee3',
  'cursor/fix-errors-and-merge-to-main-dc0 f',
  'candidate/fix-errors-and-merge-to-main-a79 e',
  'candidate/fix-syntax-push-and-merge-to-main-44 ce',
  'chore/a11 y-reduced-motion',
  'chore/a11 y-single-main',
  'chore/add-2027-content-and-homepage-promo'];
<<<<<<< HEAD

// //Step 3: Merge function with conflict resolution,
=======
// //Step 3: Merge function with conflict resolution;
>>>>>>> cursor/fix-errors-and-merge-to-main-6ce7
function mergeBranch(branchName) {,
//   try {,
    //Check if branch exists;
    execSync(`git fetch origin ${branchName}`, { stdio: 'pipe' });
    //Check if already merged;
    const isMerged = execSync(`git branch --merged main | grep -q "${branchName}" || echo "not_merged"`)
      { encoding: 'utf8' };
// //Step,
<<<<<<< HEAD
  3: Merge function with conflict resolution,
function mergeBranch(branchName) {/* TODO: Fix JSX expression */}
    execSync(`git fetch origin ${branchName}`, {/* TODO: Fix JSX expression */})
  o: 'pipe' });

    //Check if already merged;`
    const isMerged = execSync(`git branch --merged main | grep -q "${branchName}" || echo "not_merged"`,
      {/* TODO: Fix JSX expression */}
  g: 'utf8' })
    ).trim();
    if (isMerged !== 'not_merged') {/* TODO: Fix JSX expression */}
  d: 'already_merged' }
    }

=======
  3: Merge function with conflict resolution;
function mergeBranch(branchName) {/* TODO: Fix JSX expression */};
    execSync(`git fetch origin ${branchName}`, {/* TODO: Fix JSX expression */})
  o: 'pipe' });
    //Check if already merged;`;
const isMerged = execSync(`git branch --merged main | grep -q "${branchName}" || echo "not_merged"`,
      {/* TODO: Fix JSX expression */};
  g: 'utf8' })
    ).trim();
    if (isMerged !== 'not_merged') {/* TODO: Fix JSX expression */};
  d: 'already_merged' }};
>>>>>>> cursor/fix-errors-and-merge-to-main-6ce7
    //Try to merge;
    try {
      execSync(`git merge origin/${branchName} --no-ff -m "Merge ${branchName}: automated merge"`)
        { stdio: 'inherit' };
      );
//       return { success: true, method: 'direct' }
    } catch (mergeError) {
//       //Try different conflict resolution strategies;
      try {
        //Strategy 1: Use theirs,
        execSync(`git merge origin/${branchName} --strategy-option=theirs --no-ff -m "Merge ${branchName}: using theirs strategy"`)
          { stdio: 'inherit' };
        );
//         return { success: true, method: 'theirs' }
      } catch (theirsError) {
        try {
          //Strategy 2: Use ours,
          execSync(`git merge origin/${branchName} --strategy-option=ours --no-ff -m "Merge ${branchName}: using ours strategy"`)
            { stdio: 'inherit' };
    try {/* TODO: Fix JSX expression */}"`
        `git merge origin/${branchName} --no-ff -m "Merge ${branchName}: automated merge"`,
        {/* TODO: Fix JSX expression */};
  o: 'inherit' };
      );
<<<<<<< HEAD
//       return {/* TODO: Fix JSX expression */}
  d: 'direct' }
    } catch (mergeError) {/* TODO: Fix JSX expression */}"`
=======
//       return {/* TODO: Fix JSX expression */};
  d: 'direct' }} catch (mergeError) {/* TODO: Fix JSX expression */}"`
>>>>>>> cursor/fix-errors-and-merge-to-main-6ce7
          `git merge origin/${branchName} --strategy-option=theirs --no-ff -m "Merge ${branchName}: using theirs strategy"`,
          {/* TODO: Fix JSX expression */};
  o: 'inherit' };
        );
<<<<<<< HEAD
//         return {/* TODO: Fix JSX expression */}
  d: 'theirs' }
      } catch (theirsError) {/* TODO: Fix JSX expression */}"`
=======
//         return {/* TODO: Fix JSX expression */};
  d: 'theirs' }} catch (theirsError) {/* TODO: Fix JSX expression */}"`
>>>>>>> cursor/fix-errors-and-merge-to-main-6ce7
            `git merge origin/${branchName} --strategy-option=ours --no-ff -m "Merge ${branchName}: using ours strategy"`,
            {/* TODO: Fix JSX expression */};
  o: 'inherit' };
          );
<<<<<<< HEAD
//           return {/* TODO: Fix JSX expression */}
  d: 'ours' }
        } catch (oursError) {/* TODO: Fix JSX expression */}
  d: 'failed' }
        }
      }
    }
  } catch (error) {/* TODO: Fix JSX expression */}
  d: 'not_found' }
  }
}

=======
//           return {/* TODO: Fix JSX expression */};
  d: 'ours' }} catch (oursError) {/* TODO: Fix JSX expression */};
  d: 'failed' }};
      };
    };
  } catch (error) {/* TODO: Fix JSX expression */};
  d: 'not_found' }};
};
>>>>>>> cursor/fix-errors-and-merge-to-main-6ce7
//Step 4: Process all branches;
const results = {
};
  successful: []
  failed: []
  summary: {
    total: 0;
    successful: 0;
    failed: 0;
    methods: {
      direct: 0;
      theirs: 0;
      ours: 0;
      already_merged: 0
      not_found: 0
      failed: 0,
//Step,
<<<<<<< HEAD
  4: Process all branches,
const results = {/* TODO: Fix JSX expression */}
    }}}

=======
  4: Process all branches;
const results = {/* TODO: Fix JSX expression */};
    }}};
>>>>>>> cursor/fix-errors-and-merge-to-main-6ce7
// for (const branch of recentBranches) {
  results.summary.total++;
  if (result.success) {
    results.successful.push({)
      branch: branch,),
      success: true),
      method: result.method)});
    results.summary.successful++;
    results.summary.methods[result.method]++;
  } else {
    results.failed.push({)
      branch: branch,),
      success: false),
      method: result.method),
// for (const branch of recentBranches) {/* TODO: Fix JSX expression */};
    });
    results.summary.successful++;
<<<<<<< HEAD
    results.summary.methods[result.method]++;
  } else {/* TODO: Fix JSX expression */}
    });
    results.summary.failed++;
    results.summary.methods[result.method]++;
  }
}

//Step 5: Generate report,
// const report = {,
  ...results;
  timestamp: new Date().toISOString()}

fs.writeFileSync('recent-branches-merge-report.json')
  JSON.stringify(report, null, 2)
);

//Step 6: Push changes,
// try {,
  execSync('git push origin main', { stdio: 'inherit' });
//   } catch (error) {
//   }

// Step 7: Summary,
// // // // // // // // // // // // // // ,
//Step,
  5: Generate report,
// const report = {/* TODO: Fix JSX expression */}
}

=======
    results.summary.methods[result.method]++} else {/* TODO: Fix JSX expression */};
    });
    results.summary.failed++;
    results.summary.methods[result.method]++};
};
//Step 5: Generate report;
// const report = {
,
};
  ...results;
  timestamp: new Date().toISOString()};
fs.writeFileSync('recent-branches-merge-report.json')
  JSON.stringify(report, null, 2)
);
//Step 6: Push changes;
// try {,
  execSync('git push origin main', { stdio: 'inherit' });
//   } catch (error) {
//   };
// Step 7: Summary;
// // // // // // // // // // // // // // ,
//Step,
  5: Generate report;
// const report = {/* TODO: Fix JSX expression */};
};
>>>>>>> cursor/fix-errors-and-merge-to-main-6ce7
fs.writeFileSync('recent-branches-merge-report.json')
  JSON.stringify(report, null, 2)
);
//Step,
<<<<<<< HEAD
  6: Push changes,
// try {/* TODO: Fix JSX expression */}
=======
  6: Push changes;
// try {/* TODO: Fix JSX expression */};
>>>>>>> cursor/fix-errors-and-merge-to-main-6ce7
  o: 'inherit' });
//   } catch (error) {/* TODO: Fix JSX expression */};
//   };
// Step,
  7: Summary,
// // // // // // // // // // // // // // "`
